crates/bevy_ecs/src/system/builder.rs

```diff
@@ -0,0 +1,188 @@
+use std::marker::PhantomData;
+
+use bevy_utils::{all_tuples, index_tuple};
+
+use crate::{archetype::ArchetypeGeneration, prelude::World};
+
+use super::{FunctionSystem, IntoSystem, SystemMeta, SystemParam, SystemParamFunction};
+
+pub struct SystemBuilder<'w, Marker, F: SystemParamFunction<Marker>>
+where
+    Self: SystemParamBuilder<F::Param>,
+{
+    func: Option<F>,
+    builders: <Self as SystemParamBuilder<F::Param>>::Builders,
+    world: &'w mut World,
+}
+
+impl<'w, Marker: 'static, F: SystemParamFunction<Marker>> SystemBuilder<'w, Marker, F>
+where
+    Self: SystemParamBuilder<F::Param>,
+{
+    pub fn new(world: &'w mut World, func: F) -> Self {
+        Self {
+            func: Some(func),
+            builders: <Self as SystemParamBuilder<F::Param>>::Builders::default(),
+            world,
+        }
+    }
+
+    pub fn param<const I: usize>(
```

Comment:
Is there no one way to add params onto the end?

Reply:
Not that's expressible in rust's type system. It's very difficult to add a new type to a tuple, it's similar to trying to add a new type to a bundle.

The does open the door for a `DynParams` type that you can dynamically build system params inside of and then access inside the system.

Reply:
Fixed, although a `DynParams` type is still a good idea for dynamic environments that want to pass a dynamically defined set of parameters to a function.